### DISCUSSION TITLE
Mutually misaligned disk components

### DISCUSSION THREAD

**Discussion by ginski** [dullemond/radmc3d-2.0#10] [Q&A] [answered]
Hi,

I recently switched from other radiative transfer codes to radmc3d since I like the flexibility and the awesome documentation :)
However, something I don't quite understand yet is, how it might be possible to set up a model in which there are two or even multiple zones in the disk that are misaligned with respect to each other (e.g. to study shadowing in scattered light observations). Unless I am mistaken this is not demonstrated in any of the included examples. Yet I have seen publications based on radmc3d where such models are shown. 

Does an example exist that demonstrates how to do this?

Cheers
Christian

**Comment by dullemond** [comment, marked as answer]
Hi Christian, 

there is no example model, but it is indeed possible. The most straightforward way would be to start from one of the disk examples in spherical coordinates, make sure it is 3D (i.e. including the phi-coordinate), and then use a coordinate transformation from (theta,phi) to (theta',phi') to compute, for each cell, the theta' and phi' values. The r-coordinate stays the same. This way you can tilt the disk, even though the coordinate system stays fixed. If you do this transformation differently for each r, you can incorporate any warp or twist you like. The only drawback of using spherical coordinates for warped/broken disks is the coordinate singularity at the poles. If your inner disk is inclined 90 degrees w.r.t. the outer disk, the inner disk goes straight through this singularity. That can amplify coordinate effect in the model.

I think that Attila Juhasz's models of warped/broken disks are done in this way. But given my own interest in this topic, I might spend some time tomorrow or next week to create an example model that does this, and I might need you to help test it :-)

There is, however, an example called run_warpeddisk/ which does this same thing, but in cartesian coordinates. Since regular Cartesian coordinates are bad for circumstellar disks/envelopes (they have too little spatial resolution near the inner disk regions), this is done on an octree AMR grid: refining the cells hierarchically until all regions are sufficiently resolved. This works as well, but I am not sure how well it works for disk geometries. Another issue with this example model is that the model is created inside of RADMC-3D, not in the usual way in Python. That is fine, but less intuitive. 

Best wishes,
Kees

**Reply by dullemond** [reply]
Hi Christian,

I made a test model setup for warping a disk with a given unit normal vector **l**(r), i.e. you can specify for each point on the 1-D radial grid a unit vector that is normal to the disk. If you specify it everywhere as [0,0,1] (i.e. in z-direction) then the disk remains as it is. If you specify it everywhere as e.g. [1/np.sqrt(2),0,1/np.sqrt(2)] then it will be inclined toward the x-axis by 45 degrees. But if you specify it as a varying function of r (an array of dimension [nr,3], where nr is the number of grid points in radius), you can warp and twist the disk as you like.

Note: this is in spherical coordinates. This means that if you have parts of the disk very inclined (close to 90 degrees) then the disk will go straight through the coordinate singularity. The theta-grid goes not all the way exactly to the poles, but you can adjust that with `thetaup`.

You can find the model in the branch `ppdisk_warped` (git checkout ppdisk_warped) in the directory `examples/run_ppdisk_simple_warped_1/`. It does not yet have a README file, but it works the same as the other example models. Run `python problem_setup.py`, then `radmc3d mctherm`, and then go into `ipython --matplotlib` and run `plot_images.py`.

I have only tested it a bit, and the number of photon packages for the scattered light image at 1 micron is still too low. It is not an optimal model. Also the grid may need fine-tuning. The model is the warped version of the `run_ppdisk_simple_1/` model, which is really very simplified. So this warped model is only meant to show how this can be done, but a production-level model would need significant fine-tuning.

If you wish to model an "inclined disk inside a disk", you can do this with this setup. Just make sure to put a large gap in the disk, and switch from the inner ***l*** vector to the outer ***l*** vector somewhere inside this gap.

Would you be willing to test this model setup and suggest to me adjustments and improvements?

Best wishes,
Kees

**Reply by ginski** [reply]
Hi Kees,

wow thank you so much for the quick explanation and for providing an example! A summer student of mine is currently working on a model for one of the observations we recently obtained and we will very happily have a look at your example.

Cheers
Christian
